### ISSUE TITLE
A few more things to clean up in the promise land

### ISSUE THREAD

**Issue by alexcjohnson** [plotly/plotly.js#76]
- [ ] Check that we're doing things as synchronously as possible (see [comment on 70](https://github.com/plotly/plotly.js/pull/70/files#r46615093) )
- [x] Make sure _all_ plot_api calls return promises. Looks like at present `plot`, `newPlot`, `redraw`, `restyle`, and `relayout` do, but not `extendTraces`, `moveTraces` etc.

**Comment by alexcjohnson** [comment]
From a separate convo

@etpinard 
> Plotly.plot's promise story is still a bit of a mystery to me
> One thing I do know: we can't test `Plotly.plot` in jasmine without handling the `then()` regardless of the plot type

@alexcjohnson
> huh, OK, it seems weird to me. We do always return a promise, but everything should be synchronous up to that point in the normal case, so by the time the function exits it should be complete.
> might be interesting to watch [`donePlotting`](https://github.com/plotly/plotly.js/blob/master/src/plot_api/plot_api.js#L358)
and see if we are really making everything async for some dumb reason...

It's not the end of the world if plotting is less synchronous than it could be, but the experience is better for the viewer (at the very least fewer flashes of half-built plots, and likely faster display overall) if it's sync. For the developer using plotly.js it's probably no different, they should get used to expecting `Plotly.plot` and `Plotly.newPlot` (and everything else in `plot_api` but there's some more for us to clean up there) to return a promise.

**Comment by mdtusz** [comment]
@etpinard @alexcjohnson I see that the promise resolve isn't passed an argument - should we pass in the gd? That's what I would expect to be handed when using the library, and is trivial to implement because the gd is always in scope when Promise.resolve is called.

**Comment by etpinard** [comment]
@mdtusz I vote for passing the `gd`

**Comment by alexcjohnson** [comment]
@mdtusz sure, if you think it would be useful. Generally I don't see why it would, as all these functions take `gd` as an arg so the caller already has it. But I guess there could be cases, like if you bundle a bunch of plot promises together...

**Comment by mdtusz** [comment]
@alexcjohnson my main use-case is if someone does something like this:

``` javascript
// In some place, dynamically adding plot(s)
var myPlot = Plotly.plot('someDiv', data, layout);
var plots = [myPlot];


// Somewhere else in the code, maybe another file
plots.forEach(function(plot){
    plot.then(function(gd){
         gd.className = "visible";
    });
});
```

It's a bit of a contrived example, but the expected behaviour of promises is generally to receive some success data, or an error. Doesn't hurt to pass it along in any case!

**Comment by alexcjohnson** [comment]
> Doesn't hurt to pass it along in any case!

:+1: yup, that was basically the kind of case I was imagining, go for it.

**Comment by mdtusz** [comment]
https://github.com/plotly/plotly.js/pull/77

**Comment by mdtusz** [comment]
@alexcjohnson @etpinard Can we close this?

**Comment by alexcjohnson** [comment]
Ideally it would be nice to do a little exploration of whether our async behavior really is what we think it is, ie only MathJax (and maybe gl?) make plotting be async. Mainly to make it easier to write tests around plotting.

Also when it *is* async, it looks to me as though we need [`cleanUp`](https://github.com/plotly/plotly.js/blob/master/src/plot_api/plot_api.js#L338) to return `gd` as well so that the resolve arg is consistent?

**Comment by alexcjohnson** [comment]
considering this closed by #1991 and many before it
